<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>fastStructure analysis of pyRAD output &#8211; Deren Eaton</title>
<meta name="description" content="short summary here">
<meta name="keywords" content="pyrad, radseq, popgen">
<!-- Twitter Cards -->
	
		<meta name="twitter:card" content="summary_large_image">
		<meta name="twitter:image" content=
			
				"http://localhost:4000/images/header.jpg"
			
		>
	
	<meta name="twitter:title" content="fastStructure analysis of pyRAD output">
	<meta name="twitter:description" content="short summary here">
	<meta name="twitter:creator" content="@dereneaton">


<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="fastStructure analysis of pyRAD output">
<meta property="og:description" content="short summary here">
<meta property="og:url" content="http://localhost:4000/popgen/2014-04-12-faststructure-analysis-of-pyrad-output/">
<meta property="og:site_name" content="Deren Eaton">





<link rel="canonical" href="http://localhost:4000/popgen/2014-04-12-faststructure-analysis-of-pyrad-output/">
<link href="http://localhost:4000/feed.xml" type="application/atom+xml" rel="alternate" title="Deren Eaton Feed">


<!-- http://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- Google Webfonts -->
<link href='http://fonts.googleapis.com/css?family=PT+Sans+Narrow:400,700|PT+Serif:400,700,400italic' rel='stylesheet' type='text/css'>
<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.min.css">

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
	<script src="http://localhost:4000/assets/js/vendor/html5shiv.min.js"></script>
	<script src="http://localhost:4000/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://localhost:4000/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>

<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://localhost:4000/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://localhost:4000/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://localhost:4000/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://localhost:4000/images/apple-touch-icon-144x144-precomposed.png">

</head>

<body class="post">

<!--[if lt IE 9]><div class="browser-upgrade alert alert-info">You are using an <strong>outdated</strong> browser. Please <a href="http://browsehappy.com/">upgrade your browser</a> to improve your experience.</div><![endif]-->

<div class="navigation-wrapper">
	<div class="site-name">
		<a href="http://localhost:4000">Deren Eaton</a>
	</div><!-- /.site-name -->
	<div class="top-navigation">
		<nav role="navigation" id="site-nav" class="nav">
		    <ul>
		        
				<li><a href="http://localhost:4000/about/" >About</a></li>
		        
				<li><a href="http://localhost:4000/posts/" >Posts</a></li>
		        
				<li><a href="http://localhost:4000/publications/" >Publications</a></li>
		        
				<li><a href="http://localhost:4000/software/" >Software</a></li>
		        
		    </ul>
		</nav>
	</div><!-- /.top-navigation -->
</div><!-- /.navigation-wrapper -->


  <div class="image-wrap">
  <img src=
    
      "http://localhost:4000/images/header.jpg"
    
  alt="fastStructure analysis of pyRAD output feature image">
  
  </div><!-- /.image-wrap -->


<div id="main" role="main">
  <div class="article-author-side">
    
	<img src="http://localhost:4000/images/me_decorRNA_0610.jpg" class="bio-photo" alt="Deren Eaton bio photo"></a>

<h3>Deren Eaton</h3>
<p>Postdoc :: Yale University Ecology and Evolutionary Biology.</p>
<a href="http://twitter.com/dereneaton" class="author-social" target="_blank"><i class="icon-twitter"></i> Twitter</a>




<a href="http://flower-doctor.tumblr.com" class="author-social" target="_blank"><i class="icon-tumblr"></i> Tumblr</a>







  </div>
  <article>
    <div class="headline-wrap">
      <h1>fastStructure analysis of pyRAD output</h1>
    </div><!--/ .headline-wrap -->
    <div class="article-wrap">
      <h4 id="topics"><em>Topics</em></h4>

<ul>
  <li>converting <em>pyRAD</em> output into .structure format</li>
  <li>using fastStructure</li>
  <li>the effect of missing data (very relevant for RADseq)</li>
</ul>

<h4 id="structure"><em>Structure</em></h4>

<p>Jonathan Pritchard’s Bayesian clustering program <em>Structure</em> is widely used to infer population structure from genetic data. From reading the documentation it is clear there are many additional functions built into the software that are rarely used. These include setting geographic distances between populations, at least from what I’ve seen, and they bloat the software a bit, as well as the input files. </p>

<h4 id="faststructure"><em>fastStructure</em></h4>
<p>A recent paper by  (arxiv:) describes a fast new algorithm that takes advantage of () to perform <em>Structure</em> analyses much faster, and in some cases more accurately. This new version will be particularly applicable to very large SNP data sets, such as human genome data, and this drew my attention since it should be similarly applicable to SNP data sets generated with RADseq. </p>

<h4 id="getting-structure-files-from-pyrad">getting structure files from <em>pyRAD</em></h4>

<p>the program can be used for much more than its base applications, the assignment of individuals to do much more than it i</p>

<h2 id="blah">blah</h2>

      <hr />
      <footer role="contentinfo">
        <div class="article-author-bottom">
          
	<img src="http://localhost:4000/images/me_decorRNA_0610.jpg" class="bio-photo" alt="Deren Eaton bio photo"></a>

<h3>Deren Eaton</h3>
<p>Postdoc :: Yale University Ecology and Evolutionary Biology.</p>
<a href="http://twitter.com/dereneaton" class="author-social" target="_blank"><i class="icon-twitter"></i> Twitter</a>




<a href="http://flower-doctor.tumblr.com" class="author-social" target="_blank"><i class="icon-tumblr"></i> Tumblr</a>







        </div>
        <p class="byline"><strong>fastStructure analysis of pyRAD output</strong> was published on <time datetime="2014-04-15T16:51:25-04:00">April 15, 2014</time> by <a href="http://localhost:4000/about" title="About Deren Eaton">Deren Eaton</a>.</p>
      </footer>
    </div><!-- /.article-wrap -->
  </article>
</div><!-- /#main -->

<div class="footer-wrap">
  <div class="related-articles">
  <h4>You might also enjoy <small class="pull-right">(<a href="http://localhost:4000/posts/">View all posts</a>)</small></h4>
    <ul>
    
      <li><a href="http://localhost:4000/phylogenetics/2014-04-15-simulating-radseq-data-on-a-fixed-tree/" title="Simulating RADseq data on a fixed tree">Simulating RADseq data on a fixed tree</a></li>
    
      <li><a href="http://localhost:4000/web/2014-04-12-posting-with-ipython-notebook/" title="Posting with ipython notebook">Posting with ipython notebook</a></li>
    
      <li><a href="http://localhost:4000/phylogenetics/2014-04-15-inferring-branch-lengths-on-a-fixed-topology-in-beast2/" title="Inferring branch lengths on a fixed topology in BEAST2">Inferring branch lengths on a fixed topology in BEAST2</a></li>
    
    </ul>
    <hr />
  </div><!-- /.related-articles -->
  <footer>
    <span>&copy; 2014 Deren Eaton. Powered by <a href="http://jekyllrb.com">Jekyll</a> using the <a href="http://mademistakes.com/minimal-mistakes/">Minimal Mistakes</a> theme.</span>

  </footer>
</div><!-- /.footer-wrap -->

<script src="//ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://localhost:4000/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://localhost:4000/assets/js/scripts.min.js"></script>
	        

</body>
</html>