<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>Simulating RADseq data on a fixed tree &#8211; Deren Eaton</title>
<meta name="description" content="">
<meta name="keywords" content="radseq, phylogeny, popgen">
<!-- Twitter Cards -->
	
		<meta name="twitter:card" content="summary_large_image">
		<meta name="twitter:image" content=
			
				"http://localhost:4000/images/header.jpg"
			
		>
	
	<meta name="twitter:title" content="Simulating RADseq data on a fixed tree">
	<meta name="twitter:description" content="Academic Portal.">
	<meta name="twitter:creator" content="@dereneaton">


<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="Simulating RADseq data on a fixed tree">
<meta property="og:description" content="Academic Portal.">
<meta property="og:url" content="http://localhost:4000/phylogenetics/2014-04-15-simulating-radseq-data-on-a-fixed-tree/">
<meta property="og:site_name" content="Deren Eaton">





<link rel="canonical" href="http://localhost:4000/phylogenetics/2014-04-15-simulating-radseq-data-on-a-fixed-tree/">
<link href="http://localhost:4000/feed.xml" type="application/atom+xml" rel="alternate" title="Deren Eaton Feed">


<!-- http://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- Google Webfonts -->
<link href='http://fonts.googleapis.com/css?family=PT+Sans+Narrow:400,700|PT+Serif:400,700,400italic' rel='stylesheet' type='text/css'>
<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.min.css">

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
	<script src="http://localhost:4000/assets/js/vendor/html5shiv.min.js"></script>
	<script src="http://localhost:4000/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://localhost:4000/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>

<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://localhost:4000/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://localhost:4000/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://localhost:4000/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://localhost:4000/images/apple-touch-icon-144x144-precomposed.png">

</head>

<body class="post">

<!--[if lt IE 9]><div class="browser-upgrade alert alert-info">You are using an <strong>outdated</strong> browser. Please <a href="http://browsehappy.com/">upgrade your browser</a> to improve your experience.</div><![endif]-->

<div class="navigation-wrapper">
	<div class="site-name">
		<a href="http://localhost:4000">Deren Eaton</a>
	</div><!-- /.site-name -->
	<div class="top-navigation">
		<nav role="navigation" id="site-nav" class="nav">
		    <ul>
		        
				<li><a href="http://localhost:4000/about/" >About</a></li>
		        
				<li><a href="http://localhost:4000/posts/" >Posts</a></li>
		        
				<li><a href="http://localhost:4000/publications/" >Publications</a></li>
		        
				<li><a href="http://localhost:4000/software/" >Software</a></li>
		        
		    </ul>
		</nav>
	</div><!-- /.top-navigation -->
</div><!-- /.navigation-wrapper -->


  <div class="image-wrap">
  <img src=
    
      "http://localhost:4000/images/header.jpg"
    
  alt="Simulating RADseq data on a fixed tree feature image">
  
    <span class="image-credit">Photo Credit: <a href="">Deren Eaton</a></span>
  
  </div><!-- /.image-wrap -->


<div id="main" role="main">
  <div class="article-author-side">
    
	<img src="http://localhost:4000/images/me_decorRNA_0610.jpg" class="bio-photo" alt="Deren Eaton bio photo"></a>

<h3>Deren Eaton</h3>
<p>Postdoc :: Yale University Ecology and Evolutionary Biology.</p>
<a href="http://twitter.com/dereneaton" class="author-social" target="_blank"><i class="icon-twitter"></i> Twitter</a>




<a href="http://flower-doctor.tumblr.com" class="author-social" target="_blank"><i class="icon-tumblr"></i> Tumblr</a>







  </div>
  <article>
    <div class="headline-wrap">
      <h1>Simulating RADseq data on a fixed tree</h1>
    </div><!--/ .headline-wrap -->
    <div class="article-wrap">
      <h4 id="background">Background</h4>
<p>I originally developed this script for bug testing <em>pyRAD</em>, but more recently 
I’ve used similar versions of it in a number of studies to simulate RADseq data 
with varying amounts of divergence between samples, proportions of missing data, and with or 
without gene flow between lineages. I figured it may be useful to others so I am making 
it available here, with instructions. </p>

<p>The script sets up code to call the Python module <a href="http://http://egglib.sourceforge.net/">egglib</a>
to simulate sequence data under a coalescent model, and then it dresses up the sequences
to act as if they had been generated as a reduced representation genomic library (RRL). So 
you will need to first install the egglib package found at the link above. </p>

<h4 id="calling-the-script">Calling the script</h4>
<p>simRLLs.py takes five arguments:</p>

<ol>
  <li>Allow mutations to restriction sites (1/0).</li>
  <li>Allow indels (decimal). Probability a mutation is a deletion.</li>
  <li>Number of individuals per tip taxon (int; minimum 1).</li>
  <li>Number of loci to sample (int; multiple of 100 or 1000)</li>
  <li>Data type (string; options: rad, ddrad, gbs, pairddrad, pairgbs)</li>
  <li>Raw or Assembled data (0/1)</li>
  <li>Prefix name for output files (string)</li>
</ol>

<div class="highlight"><pre><code class="bash">python simRRLs.py 0 0.1 10 rad 1 simrads
</code></pre></div>

<p>Ideally one would be able to put in any topology on which to evolve the sequences, but 
I haven’t yet had time to work this into the script. For now sequences are evolved only
on the fixed tree shown below: </p>

<h4 id="insert-image-here">insert image here.</h4>

<h5 id="option-1">option 1</h5>
<p>The outgroup taxon “X” is not included in the output data, but is used in the simulation 
to polarize mutations relative to an outgroup. If option 1 is set to allow mutations to 
restriction sites then any mutations in restriction sites relative to the outgroup will 
cause reads to be excluded from the library. </p>

<h5 id="option-2">option 2</h5>
<p>Any point mutations relative to the outgroup taxon “X” may be converted to a deletion
with the probability set here. If 0, no indels are present in the data set. </p>

<h5 id="option-5">option 5</h5>

<h2 id="output">Output</h2>
<p>This will create two output files. If the prefix name you entered was “simrads” you will 
get back simrads.fastq.gz and simarads.barcodes. The first is a compressed fastq file with
the sequence data and the latter is a text file mapping barcodes to sample names. </p>


      <hr />
      <footer role="contentinfo">
        <div class="article-author-bottom">
          
	<img src="http://localhost:4000/images/me_decorRNA_0610.jpg" class="bio-photo" alt="Deren Eaton bio photo"></a>

<h3>Deren Eaton</h3>
<p>Postdoc :: Yale University Ecology and Evolutionary Biology.</p>
<a href="http://twitter.com/dereneaton" class="author-social" target="_blank"><i class="icon-twitter"></i> Twitter</a>




<a href="http://flower-doctor.tumblr.com" class="author-social" target="_blank"><i class="icon-tumblr"></i> Tumblr</a>







        </div>
        <p class="byline"><strong>Simulating RADseq data on a fixed tree</strong> was published on <time datetime="2014-04-15T17:56:47-04:00">April 15, 2014</time> and last modified on <time datetime="2014-04-15">April 15, 2014</time> by <a href="http://localhost:4000/about" title="About Deren Eaton">Deren Eaton</a>.</p>
      </footer>
    </div><!-- /.article-wrap -->
  </article>
</div><!-- /#main -->

<div class="footer-wrap">
  <div class="related-articles">
  <h4>You might also enjoy <small class="pull-right">(<a href="http://localhost:4000/posts/">View all posts</a>)</small></h4>
    <ul>
    
      <li><a href="http://localhost:4000/popgen/2014-04-12-faststructure-analysis-of-pyrad-output/" title="fastStructure analysis of pyRAD output">fastStructure analysis of pyRAD output</a></li>
    
      <li><a href="http://localhost:4000/web/2014-04-12-posting-with-ipython-notebook/" title="Posting with ipython notebook">Posting with ipython notebook</a></li>
    
      <li><a href="http://localhost:4000/phylogenetics/2014-04-15-inferring-branch-lengths-on-a-fixed-topology-in-beast2/" title="Inferring branch lengths on a fixed topology in BEAST2">Inferring branch lengths on a fixed topology in BEAST2</a></li>
    
    </ul>
    <hr />
  </div><!-- /.related-articles -->
  <footer>
    <span>&copy; 2014 Deren Eaton. Powered by <a href="http://jekyllrb.com">Jekyll</a> using the <a href="http://mademistakes.com/minimal-mistakes/">Minimal Mistakes</a> theme.</span>

  </footer>
</div><!-- /.footer-wrap -->

<script src="//ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://localhost:4000/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://localhost:4000/assets/js/scripts.min.js"></script>
	        

</body>
</html>