---
layout: page
title: "Vibgrant"
date: 
modified:
excerpt:
tags: []
image: 
  feature: header.png
  credit: Deren Eaton
comments: true
---

[Figure_1](Figure_1_leaves.png) -- contrasting leaf forms, sems, hairs, histograms.  

[Figure_2 map and hypotheses](Figure_2_map.svg) ([PDF](Figure_2_map.pdf)) ([PNG](Figure_2_map.png))-- A map showing the two original divergence hypotheses. Orange indicates areas where we will focus our collecting efforts, green area will be examined in the context of phylogeny using samples from herbaria.  

[Figure_3 heteroblasty](Figure_3_hetero.png) : Heteroblasty in Viburnun, microcarpum, and Spriggs figure.   

[Figure_4 hybrids](Figure_4_hybrid.svg) ([PDF](Figure_4_hybrid.pdf)) ([PNG](Figure_4_hybrid.png)): Recombinant phenotypes seen in hybrids.  

[Figure_5 Ellipses in theory](Figure_5_Ellipsetheory.svg) ([PDF](Figure_5_Ellipsetheory.pdf)) ([PNG](Figure_5_Ellipsetheory.png)) : Expectations and explanation of VCV matrices.  

[Figure_6 RADseq tree & ABBA-BABA](Figure_6_radtree.svg) ([PDF](Figure_6_radtree.pdf)) ([PNG](Figure_6_radtree.png) -- Our RADseq phylogeny for 16 species in _Oreinotinus_. The tree supports replicate parallel leaf divergences. (B) Morphological hybrids sampled where sister-species co-occur show genomic evidence of admixture (ABBA>BABA). (C) Introgression between lineages from adjacent regions is not supported, which provides further support that our phylogeny is correct and does not simply reflect recent gene flow.  

[Figure_7 Cross section environment](Figure_7_cross.pdf)) ([EPS](Figure_7_cross.eps)) Cross-section through Oaxaca and Chiapas with PCA environment and aerial light capture.  

[Figure_8] -- Niche models.  

[Figure_9] -- Empirical G/D matrix, and variation.  

[Figure_10] -- seedlings.  

[Figure_11] -- Timeline.  


------------------------------------------   



[Development line plots (main pops)](Full_D.html), ([Full_D_alt.html](other pops)) : Developmental variation in jucundum (PN) and lautum (Teo; subsampled first 300 branches). Each point represent the mean leaf morphology in the first leaf pair (V1) of a branch and a vector is drawn from there to the mean leaf morphology at the last leaf pair sampled in that branch.  

[Across-individual V1 line plots (all pops)](Full_I1.html), ([V2s](Full_I2.html)) : Here lines connect mean V1 leaf values across branches from the same individual. The lines thus show variation within individuals at developmentally homologous leaves (devologues?) and any apparent pattern of parallel lines across branches would represent the across-individual direction of variation. The third panel shows the mean V1 values of each individual, and thus the across-individual patterns.The V2s show the same across-individual variation as V1s!


[Fig. 15 -- Trait histograms](Full_H.html) : Histograms showing diffrences in leaf trait distributions between species.  

[Scatterplots](Full_S.html) : Bivariate scatterplots showing the distribution of leaf trait values between species.  

[Scatterplots x Pop](Full_S_pops.html) : The same as above but with each population highlighted.  

[Scatterplots x Leaf set](Full_S_tiers.html) : The same as above but with each leaf tier highlighted.  

[Fig. 6 -- Empirical ellipse]() : Coming soon...


[Ellipse demo](Ellipse_demo.html)  
[Ellipse_demo2](Ellipse_demo2.html)  
[Ellipse_test](Ellipse_testing.html)









