<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>Simulating RADseq data on a fixed tree &#8211; Deren Eaton</title>
<meta name="description" content="a script using Egglib">
<meta name="keywords" content="radseq, phylogeny, popgen">
<!-- Twitter Cards -->
	
		<meta name="twitter:card" content="summary_large_image">
		<meta name="twitter:image" content=
			
				"http://dereneaton.github.io/images/header.jpg"
			
		>
	
	<meta name="twitter:title" content="Simulating RADseq data on a fixed tree">
	<meta name="twitter:description" content="a script using Egglib">
	<meta name="twitter:creator" content="@dereneaton">


<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="Simulating RADseq data on a fixed tree">
<meta property="og:description" content="a script using Egglib">
<meta property="og:url" content="http://dereneaton.github.io/phylogenetics/simulating-radseq-data-on-a-fixed-tree/">
<meta property="og:site_name" content="Deren Eaton">





<link rel="canonical" href="http://dereneaton.github.io/phylogenetics/simulating-radseq-data-on-a-fixed-tree/">
<link href="http://dereneaton.github.io/feed.xml" type="application/atom+xml" rel="alternate" title="Deren Eaton Feed">


<!-- http://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- Google Webfonts -->
<link href='http://fonts.googleapis.com/css?family=PT+Sans+Narrow:400,700|PT+Serif:400,700,400italic' rel='stylesheet' type='text/css'>
<!-- For all browsers -->
<link rel="stylesheet" href="http://dereneaton.github.io/assets/css/main.min.css">

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
	<script src="http://dereneaton.github.io/assets/js/vendor/html5shiv.min.js"></script>
	<script src="http://dereneaton.github.io/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://dereneaton.github.io/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>

<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://dereneaton.github.io/images/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://dereneaton.github.io/images/favicon.ico">
<!-- 48x48 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://dereneaton.github.io/images/favicon.ico">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://dereneaton.github.io/images/favicon.ico">
<!-- 128x128 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="128x128" href="http://dereneaton.github.io/images/favicon.ico">



</head>

<body class="post">

<!--[if lt IE 9]><div class="browser-upgrade alert alert-info">You are using an <strong>outdated</strong> browser. Please <a href="http://browsehappy.com/">upgrade your browser</a> to improve your experience.</div><![endif]-->

<div class="navigation-wrapper">
	<div class="site-name">
		<a href="http://dereneaton.github.io">Deren Eaton</a>
	</div><!-- /.site-name -->
	<div class="top-navigation">
		<nav role="navigation" id="site-nav" class="nav">
		    <ul>
		        
				<li><a href="http://dereneaton.github.io/about/" >About</a></li>
		        
				<li><a href="http://dereneaton.github.io/posts/" >Posts</a></li>
		        
				<li><a href="http://dereneaton.github.io/publications/" >Publications</a></li>
		        
				<li><a href="http://dereneaton.github.io/software/" >Software</a></li>
		        
		    </ul>
		</nav>
	</div><!-- /.top-navigation -->
</div><!-- /.navigation-wrapper -->


  <div class="image-wrap">
  <img src=
    
      "http://dereneaton.github.io/images/header.jpg"
    
  alt="Simulating RADseq data on a fixed tree feature image">
  
    <span class="image-credit">Photo Credit: <a href="">Deren Eaton</a></span>
  
  </div><!-- /.image-wrap -->


<div id="main" role="main">
  <div class="article-author-side">
    
	<img src="http://dereneaton.github.io/images/me_decorRNA_0610.jpg" class="bio-photo" alt="Deren Eaton bio photo"></a>

<h3>Deren Eaton</h3>
<p>Postdoc :: Yale University Ecology and Evolutionary Biology.</p>
<a href="http://twitter.com/dereneaton" class="author-social" target="_blank"><i class="icon-twitter"></i> Twitter</a>




<a href="http://flower-doctor.tumblr.com" class="author-social" target="_blank"><i class="icon-tumblr"></i> Tumblr</a>







  </div>
  <article>
    <div class="headline-wrap">
      <h1>Simulating RADseq data on a fixed tree</h1>
    </div><!--/ .headline-wrap -->
    <div class="article-wrap">
      <hr />

<h3 id="download-the-script">Download the script</h3>
<ul>
  <li><a href="/downloads/simRLLs.py">simRRLs.py</a></li>
</ul>

<hr />

<h3 id="background">Background</h3>
<p>I originally developed this script to simulate data for bug testing <em>pyRAD</em>,
but more recently I’ve used several variants of it to simulate data 
with different amounts of divergence between samples, proportions of missing data,
and with or without gene flow between lineages. 
I figured it may be useful to others so I am making 
it available here, with some instruction.</p>

<p>The script sets up a species tree and simulates sequence data on this tree under a 
coalescent model, then dresses up the data to act as if it had been generated as a
reduced representation genomic library (RRL), such as RADseq, ddRAD or GBS. 
To use the script you will need to first install the excellent <a href="http://egglib.sourceforge.net/"><em>Egglib</em> 
Python package</a>, 
used here to perform the coalescent simulations, and it also requires
the common Python package <em>Numpy</em>.</p>

<h3 id="calling-the-script">Calling the script</h3>
<p><em>simRLLs.py</em> takes seven arguments, some of which are further described below:</p>

<ol>
  <li>Allow indels (decimal). Probability a mutation is a deletion.</li>
  <li>Allow mutations to restriction sites (1/0).</li>
  <li>Number of loci to sample (int; multiple of 100 or 1000)</li>
  <li>Number of individuals per tip taxon (int; minimum 1).</li>
  <li>Size selection of fragments (int,int)</li>
  <li>Data type (string; options: rad, ddrad, gbs, pairddrad, pairgbs)</li>
  <li>Prefix name for output files (string)</li>
</ol>

<h4 id="sequences-are-evolved-on-this-12-taxon-species-tree">Sequences are evolved on this 12 taxon species tree</h4>
<p>Ideally one would be able to pass as an argument any topology on which to evolve the sequences,
but for now there is simply one topology hard-coded into the script. 
This can easily be changed by editing the script.<br />
The outgroup taxon “X” is not included in the output data, but is used in the simulation 
to polarize mutations relative to an outgroup. </p>

<p><img src="/images/setupsims.png" alt="simtreeimage" /></p>

<h5 id="arg-1">arg 1</h5>
<p>Any point mutations relative to the outgroup taxon “X” will be converted to a deletion
with the probability set here. If 0, no indels are present in the data set. </p>

<h5 id="arg-2">arg 2</h5>
<p>If this option is set to allow mutations in the restriction cut site,
then any mutations arising in this region relative to the outgroup will 
cause reads to be excluded from the library. This type of “locus dropout” is commonly 
considered a problem with empirical RADseq data. By simulating sequences this way they 
will have missing data that is phylogenetically correlated similar to empirical data sets. </p>

<h5 id="arg-5">arg 5</h5>
<p>This is mainly used for developing filtering/trimming options in <em>pyRAD</em>. If the fragment
length is very small relative to the sequence length then sequences will include 
Illumina adapters, and some reads (GBS or paired-end) can include the barcode sequences
where two reads overshoot each other in the reverse direction. As long as the lower range
is higher than 200 this has no effect on the results. Fragment lengths are randomly selected
from within the size selection window. </p>

<h5 id="arg-6">arg 6</h5>
<p>Locus dropout will differ between RAD and the other methods since RAD data have 
a restriction cut site on only one side of sequences, whereas the other library
preparation methods have restriction sites on both ends of sequences, and thus greater
probability (assuming cut sites occur with equal frequency) of locus dropout. </p>

<h4 id="calling-the-script-1">Calling the script</h4>
<div><div class="CodeRay">
  <div class="code"><pre>python simRRLs.py 0 0.1 10 400,800 rad 1 simrads
</pre></div>
</div>
</div>

<h4 id="output">Output</h4>
<p>Two data files are created, for our example these are <code>simrads.fastq.gz</code> and <code>simrads.barcodes</code>.
The first is a compressed fastq file with sequence data and the latter is a text file mapping 
barcodes to sample names. If you select a paired-end data type it will create two sequence 
files, one with “R1” in the name and the other with “R2”. With the data and a barcode map
the data can then be analyzed in either <em>pyRAD</em> or a similar program such as <em>stacks</em>. </p>

<h4 id="adding-introgression-between-lineages">Adding introgression between lineages</h4>
<p>Adding a line of code in the script like below can introduce gene flow
between different lineages of the tree. In this case we will
set that migration occurs between time=0.5 and time=1.0
from taxon number 3 (named “1C”) into taxon number 2 (named “1B”), 
with a frequency of 4*N*0.001 migrants per generation. 
Remember time is counted backwards from the present (coalescent),
so migration is zero at time=0, then is starts at time=0.5,
and we set gene flow back to zero at time=1.0. </p>

<div><div class="CodeRay">
  <div class="code"><pre>paramSet.changePairwiseMigrationRate(<span class="float">0.5</span>, <span class="integer">2</span>, <span class="integer">3</span>, <span class="integer">4</span>*N*<span class="float">0.001</span>)
paramSet.changePairwiseMigrationRate(<span class="float">1.0</span>, <span class="integer">2</span>, <span class="integer">3</span>, <span class="float">0.0</span>)
</pre></div>
</div>
</div>


      <hr />
      <footer role="contentinfo">
        <div class="article-author-bottom">
          
	<img src="http://dereneaton.github.io/images/me_decorRNA_0610.jpg" class="bio-photo" alt="Deren Eaton bio photo"></a>

<h3>Deren Eaton</h3>
<p>Postdoc :: Yale University Ecology and Evolutionary Biology.</p>
<a href="http://twitter.com/dereneaton" class="author-social" target="_blank"><i class="icon-twitter"></i> Twitter</a>




<a href="http://flower-doctor.tumblr.com" class="author-social" target="_blank"><i class="icon-tumblr"></i> Tumblr</a>







        </div>
        <p class="byline"><strong>Simulating RADseq data on a fixed tree</strong> was published on <time datetime="2014-04-18T00:00:00-04:00">April 18, 2014</time> and last modified on <time datetime="2014-04-15">April 15, 2014</time> by <a href="http://dereneaton.github.io/about" title="About Deren Eaton">Deren Eaton</a>.</p>
      </footer>
    </div><!-- /.article-wrap -->
  </article>
</div><!-- /#main -->

<div class="footer-wrap">
  <div class="related-articles">
  <h4>You might also enjoy <small class="pull-right">(<a href="http://dereneaton.github.io/posts/">View all posts</a>)</small></h4>
    <ul>
    
      <li><a href="http://dereneaton.github.io/web/new-jekyll-website/" title="New static Jekyll website">New static Jekyll website</a></li>
    
    </ul>
    <hr />
  </div><!-- /.related-articles -->
  <footer>
    <span>&copy; 2014 Deren Eaton. Powered by <a href="http://jekyllrb.com">Jekyll</a> using the <a href="http://mademistakes.com/minimal-mistakes/">Minimal Mistakes</a> theme.</span>

  </footer>
</div><!-- /.footer-wrap -->

<script src="//ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://dereneaton.github.io/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://dereneaton.github.io/assets/js/scripts.min.js"></script>

<!-- Asynchronous Google Analytics snippet -->
<script>
  var _gaq = _gaq || [];
  var pluginUrl = 
 '//www.google-analytics.com/plugins/ga/inpage_linkid.js';
  _gaq.push(['_require', 'inpage_linkid', pluginUrl]);
  _gaq.push(['_setAccount', 'UA-50125500-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>
	        

</body>
</html>